src/emitter/json.rs

```diff
@@ -0,0 +1,152 @@
+use super::*;
+use crate::rustfmt_diff::{make_diff, DiffLine, Mismatch};
+use serde::Serialize;
+use serde_json::to_string as to_json_string;
+use std::io::{self, Write};
+use std::path::Path;
+
+#[derive(Debug, Default)]
+pub(crate) struct JsonEmitter;
+
+#[derive(Debug, Default, Serialize)]
+struct MismatchedBlock {
+    original_begin_line: u32,
+    original_end_line: u32,
+    expected_begin_line: u32,
+    expected_end_line: u32,
+    original: String,
+    expected: String,
+}
+
+#[derive(Debug, Default, Serialize)]
+struct MismatchedFile {
+    name: String,
+    mismatches: Vec<MismatchedBlock>,
+}
+
+impl Emitter for JsonEmitter {
+    fn emit_header(&self, output: &mut dyn Write) -> Result<(), io::Error> {
+        write!(output, "[")?;
+        Ok(())
+    }
+
+    fn emit_footer(&self, output: &mut dyn Write, has_diff: bool) -> Result<(), io::Error> {
+        let prefix = if has_diff { "\u{8}" } else { "" };
```

Comment:
Since each file is processed independently, each file that has mismatches is serialized independently. The `emit_header` and `emit_footer` then collect all of those into an array, sample below pretty printed for ease of reading:

```json
[ 
  {
    "name": "file1.rs",
    "mismatches": [...],
  },
  {
    "name": "file2.rs",
    "mismatches": [...],
  }
]
```

a `,` suffix is added after each serialized file to ensure the array is valid. However, this does result in a trailing `,` after the last file which would be invalid JSON:
```json
[ {...}, {...}, ]
```

This is adding a backspace to remove that last trailing `,` before inserting the closing `]`, but only does so if there were any files with diffs. That way if there were no formatting issues, then the result is `[]` and there's no need to include a backspace

Reply:
Actually after some more thought, I don't believe this approach will suffice. It works fine when printing directly to stdout, but if the output is piped elsewhere (for example to a file, which seems a likely scenario for json output) then the backspace approach does not work.